src/lib/kcps-budget-manager.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Krzysztof Frydryk <frydryk.krzysztof@intel.com>
+
+#include <rtos/spinlock.h>
+#include <sof/sof.h>
+#include <stdint.h>
+#include <sof/lib/kcps-budget-manager.h>
+#include <rtos/clk.h>
+#include <sof/sof.h>
+#include <sof/math/numbers.h>
+#ifdef __ZEPHYR__
+#include <zephyr/sys/util.h>
+#endif /* __ZEPHYR__ */
+
+static struct kcps_budget_data kcps_data;
+
+static int request_freq_change(int core, int freq)
+{
+	int current_freq;
+	int selected_freq_id = 0;
+	struct clock_info *clk;
+
+	clk = clocks_get() + core;
+
+	for (int i = 0; i < clk->freqs_num; i++) {
+		if (freq < clk->freqs[i].freq) {
+			selected_freq_id = i;
+			break;
+		}
+	}
+
+	/* don't change clock frequency if already using proper clock */
+	current_freq = clock_get_freq(core);
+	if (clk->freqs[selected_freq_id].freq != current_freq)
+		clock_set_freq(core, freq);
+
+	return 0;
+}
+
+static int max_core_consumption(void)
+{
+	int result = 0;
+
+	for (int i = 0; i < CONFIG_CORE_COUNT; i++)
+		result = MAX(result, kcps_data.kcps_consumtion[i]);
+
+	return result;
+}
+
+int core_kcps_adjust(int core, int kcps_delta)
+{
+	k_spinlock_key_t key;
+	int freq;
+
+	key = k_spin_lock(&kcps_data.lock);
+	kcps_data.kcps_consumtion[core] += kcps_delta;
+
+	/* set clock according to maximum requested mcps budget */
+	freq = max_core_consumption();
+
+	for (int i = 0; i < CONFIG_CORE_COUNT; i++)
+		request_freq_change(i, freq);
```

Comment:
Shouldn't we only modify the frequency of the core this request is for?

Reply:
This might not be generally true, but currently on all supported platforms the `core` information is ignored when actually setting cpu clock, meaning there is one selection for all cores. So I suppose we'd like to inform all cores (`clock_set_freq` sends notifications about clock change). Alternatively I can make this behavior platform specific.